framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -185,13 +190,80 @@ private AnnotationMirror createNumberAnnotationMirror(List<Number> values) {
-                        boolVal, result.getResultValue().getUnderlyingType());
-        return new RegularTransferResult<>(newResultValue, result.getRegularStore());
+            CFStore thenStore,
+            CFStore elseStore,
+            List<Boolean> resultValues,
+            TypeMirror underlyingType) {
+        AnnotationMirror boolVal = atypefactory.createBooleanAnnotation(resultValues);
+        CFValue newResultValue = analysis.createSingleAnnotationValue(boolVal, underlyingType);
+        if (elseStore != null) {
+            return new ConditionalTransferResult<>(newResultValue, thenStore, elseStore);
+        } else {
+            return new RegularTransferResult<>(newResultValue, thenStore);
+        }
+    }
+
+    @Override
+    public TransferResult<CFValue, CFStore> visitFieldAccess(
+            FieldAccessNode node, TransferInput<CFValue, CFStore> in) {
+
+        TransferResult<CFValue, CFStore> result = super.visitFieldAccess(node, in);
+
+        modifyArrayLenBasedOnIntValOnArrayLength(node, result.getRegularStore());
+
+        return result;
+    }
+
+    private void modifyArrayLenBasedOnIntValOnArrayLength(
+            FieldAccessNode arrayLengthNode, CFStore store) {
+        // If array.length is encountered, transform its @IntVal annotation into an @ArrayLen annotation for array.
+        if (!NodeUtils.isArrayLengthFieldAccess(arrayLengthNode)) {
+            return;
+        }
+        CFValue value =
+                store.getValue(
+                        FlowExpressions.internalReprOf(analysis.getTypeFactory(), arrayLengthNode));
+        if (value == null) {
+            return;
+        }
+        AnnotationMirror lengthAnno =
+                AnnotationUtils.getAnnotationByClass(value.getAnnotations(), IntVal.class);
+
+        if (lengthAnno != null) {
+            List<Long> lengthValues = ValueAnnotatedTypeFactory.getIntValues(lengthAnno);
+            List<Integer> arrayLenValues = new ArrayList<>(lengthValues.size());
+            for (Long l : lengthValues) {
+                arrayLenValues.add(l.intValue());
+            }
+            AnnotationMirror newArrayAnno = atypefactory.createArrayLenAnnotation(arrayLenValues);
+            AnnotationMirror oldArrayAnno =
+                    atypefactory.getAnnotationMirror(
+                            arrayLengthNode.getReceiver().getTree(), ArrayLen.class);
+            AnnotationMirror combinedAnno;
+            if (oldArrayAnno == null) {
+                combinedAnno = newArrayAnno;
+            } else {
+                combinedAnno =
+                        atypefactory
+                                .getQualifierHierarchy()
+                                .greatestLowerBound(oldArrayAnno, newArrayAnno);
+            }
+            Receiver arrayRec =
+                    FlowExpressions.internalReprOf(
+                            analysis.getTypeFactory(), arrayLengthNode.getReceiver());
+            store.clearValue(arrayRec);
```

Comment:
Why clear the old value?  Isn't combinedAnno always a subtype of the previous annotation?